litestar/plugins/problem_details.py

```diff
@@ -0,0 +1,111 @@
+
+if TYPE_CHECKING:
+    from litestar.config.app import AppConfig
+
+ProblemDetailsExceptionT = TypeVar("ProblemDetailsExceptionT", bound="ProblemDetailsException")
+ProblemDetailsExceptionHandlerType = Callable[[Request, ProblemDetailsExceptionT], Response]
+ExceptionToProblemDetailMapType = Mapping[type[ExceptionT], Callable[[ExceptionT], ProblemDetailsExceptionT]]
+
+
+def _problem_details_exception_handler(request: Request, exc: ProblemDetailsException) -> Response:
+    return exc.to_response(request)
+
+
+def _create_exception_handler(
+    exc_to_problem_details_exc_fn: Callable[[ExceptionT], ProblemDetailsException], exc_type: type[ExceptionT]
+) -> ExceptionHandler:
+    def _exception_handler(req: Request, exc: exc_type) -> Response:
+        problem_details_exc = exc_to_problem_details_exc_fn(exc)
+
+        return problem_details_exc.to_response(req)
+
+    return _exception_handler
+
+
+class ProblemDetailsException(HTTPException):
+    _PROBLEM_DETAIL_HEADER = "application/problem+json"
+
+    def __init__(
+        self,
+        type_: str = "about:blank",
+        status_code: int | None = None,
+        title: str | None = None,
+        instance: str | None = None,
+        detail: str | dict[str, Any] | list[Any] | None = None,
+        extra: dict[str, Any] | None = None,
+        headers: dict[str, Any] | None = None,
+    ) -> None:
+        self.type_ = type_
+        self.status_code = status_code or self.status_code
+        self.title = title
+        self.instance = instance
+        self.detail = detail  # type: ignore
+        self.extra = extra
+        self.headers = headers
+
+    def to_response(self, request: Request) -> Response:
+        problem_details = {"type": self.type_, "status": self.status_code}
+        if self.title is not None:
+            problem_details["title"] = self.title
+        if self.instance is not None:
+            problem_details["instance"] = self.instance
+        if self.detail is not None:
+            problem_details["detail"] = self.detail
+        if self.extra is not None:
+            problem_details.update(self.extra)
+
+        content = encode_json(problem_details, request.route_handler.default_serializer)
+        headers = self.headers or {}
+        headers["content-type"] = self._PROBLEM_DETAIL_HEADER
+
+        return Response(content, headers=headers)
+
+    @classmethod
+    def from_http_exception(cls, exc: HTTPException) -> ProblemDetailsException:
+        return ProblemDetailsException(status_code=exc.status_code)
+
+
+@dataclass
+class ProblemDetailsConfig:
+    exception_handler: ProblemDetailsExceptionHandlerType = _problem_details_exception_handler
+    """The exception handler used for ``ProblemdetailsException.``"""
+
+    enable_for_all_http_exceptions: bool = False
+    """Flag indicating whether to convert all exceptions into ``ProblemDetailsException.``"""
+
+    exception_to_problem_detail_map: ExceptionToProblemDetailMapType | None = None
```

Comment:
Is there any reason or value in this being `None`? I am thinking it is better to make the attribute non-optional and fallback to empty dict `{}`. It will also make static type checkers happier as it avoids the need for `None` checks before each read or write to it.

Reply:
Oh you're thinking of instantiating the config and then adding to the map? I'd just naively assumed that the usage would something like the following:

```python

config = ProblemDetailsConfig(exception_to_problem_detail_map=my_custom_dictionary_instance)
```

I can see that people may prefer to modify the attribute on the config object directly, so I'm fine with defaulting to an empty dictionary. I'd just kept it as `None` as the default to semantically indicate that the value is optional.